Program.cs

```diff
@@ -470,66 +471,125 @@ static P2pEndPoint GetExternalEndPoint(Socket socket)
                     Internal = (socket.LocalEndPoint as IPEndPoint)
                 };
             }
-
+            
             Console.WriteLine("Could not find a working STUN server");
+
             return null;
         }
 
-
         static int SleepTime(DateTime now)
         {
-            List<int> seconds = new List<int>() {10, 20, 30, 40, 50, 60};
+            List<int> seconds = new List<int>() { 10, 20, 30, 40, 50, 60 };
 
             int next = seconds.Find(x => x > now.Second);
 
             return next - now.Second;
         }
 
-        static UdtSocket PeerConnect(Socket socket, string remoteAddr, int remotePort)
+        static async Task<UdtSocket> PeerConnectAsync(string remoteAddress, int[] remotePorts)
         {
-            bool bConnected = false;
-            int retry = 0;
+            if (!sockets.Any())
+                return null;
 
-            UdtSocket client = null;
+            Console.WriteLine("Trying to connect to peer on different ports");
 
-            while (!bConnected)
+            UdtSocket udtSocket;
+            CancellationTokenSource cts;
+            var retry = 0;
+
+            do
             {
-                try
-                {
-                    DateTime now = InternetTime.Get();
+                Console.WriteLine($"Try {retry++}:");
+                Console.WriteLine();
+
+                cts = new CancellationTokenSource();
 
-                    int sleepTimeToSync = SleepTime(now);
+                var taskList = remotePorts.Select((port, index) => ConnectAsync(sockets[index], remoteAddress, port, cts.Token)).ToList();
 
-                    Console.WriteLine("[{0}] - Waiting {1} sec to sync with other peer",
-                        now.ToLongTimeString(),
-                        sleepTimeToSync);
-                    System.Threading.Thread.Sleep(sleepTimeToSync * 1000);
+                var now = InternetTime.Get();
+                var sleepTimeToSync = SleepTime(now);
 
-                    GetExternalEndPoint(socket);
+                Console.WriteLine("[{0}] - Waiting {1} sec to sync with other peer", now.ToLongTimeString(), sleepTimeToSync);
 
-                    if (client != null)
-                        client.Close();
+                await Task.Delay(TimeSpan.FromSeconds(sleepTimeToSync), cts.Token);
 
-                    client = new UdtSocket(socket.AddressFamily, socket.SocketType);
+                Console.WriteLine("Please wait... ");
+                udtSocket = await (await Task.WhenAny(taskList));
+            } while (udtSocket == null);

```

Comment:
how could udtSocket ever be null?

Reply:
In the ConnectAsync method if an exception has thrown, thanks will be returned

Reply:
but then shouldn't the do..while be instead in the task that calls ConnectAsync?

Reply:
otherwise, it could happen that the fastest task to execute is the one that raises the exception and receives null, which then makes the WhenAny select the faulty one, which doesn't seem to be the best outcome

Reply:
If the socket fails to establish a connection, it will take several seconds for an exception to be thrown (after a timeout) and the null to be returned. If the first null is returned, the connecting process is performed again. If the first socket succeeds in connecting, the socket value will be returned and the rest of the tasks will be canceled.

Reply:
>  it will take several seconds for an exception to be thrown (after a timeout) and the null to be returned

That's a very big assumption

> If the first null is returned, the connecting process is performed again.

Launch the 10 threads/tasks again, when the other 9 are already running and well? that seems very efficient.

I'm thinking that if you don't capture the exception to return null (so instead let it throw), then the worst that will happen is that Task.WhenAny() will not take in account that task (because it didn't succeed). This could be more efficient.

Reply:
If I don't capture the exception, the app will stop running. The task will not necessarily run on the new thread, and the main thread of the app may run the tasks.
![image](https://user-images.githubusercontent.com/1659032/107749982-99bb9f80-6d30-11eb-9a78-c480113e52f2.png)

`WhenAny` can be replaced with `WhenAll` but if you run the app you will get noticed that if a socket fails to connect it takes several seconds for an exception to be thrown, while another socket connected successfully and has to wait for the result of sockets.

Reply:
Any ideas?

Reply:
WhenAll would defeat the purpose of this change.

Let me think about this today and test a couple of things.

Reply:
Sorry for the delay Moshen! How about if we use this? https://stackoverflow.com/a/49313971/544947

Reply:
I made the changes، please check out [1acbc3f](https://github.com/nblockchain/p2pcopy/pull/2/commits/1acbc3fcfb1337e386aa612cee910f4113113470).

Reply:
looks great, I'll test it today or tomorrow